src/main/java/com/iota/iri/service/milestone/MilestoneSolidifier.java

```diff
@@ -0,0 +1,339 @@
+     * be updated without iterating over all entries. If the queue reached its capacity already, we replace the entry
+     * marked by the {@link #youngestMilestoneInQueue} marker and update the marker by recalculating it using
+     * {@link #determineYoungestMilestoneInQueue()}.
+     *
+     * @param milestoneEntry entry from the {@link #unsolidMilestonesPool} that shall get added to the queue
+     */
+    private void addToSolidificationQueue(Map.Entry<Hash, Integer> milestoneEntry) {
+        if (milestonesToSolidify.containsKey(milestoneEntry.getKey())) {
+            return;
+        }
+
+        if (milestonesToSolidify.size() < SOLIDIFICATION_QUEUE_SIZE) {
+            milestonesToSolidify.put(milestoneEntry.getKey(), milestoneEntry.getValue());
+
+            if (youngestMilestoneInQueue == null || milestoneEntry.getValue() > youngestMilestoneInQueue.getValue()) {
+                youngestMilestoneInQueue = milestoneEntry;
+            }
+        } else if (milestoneEntry.getValue() < youngestMilestoneInQueue.getValue()) {
+            milestonesToSolidify.remove(youngestMilestoneInQueue.getKey());
+            milestonesToSolidify.put(milestoneEntry.getKey(), milestoneEntry.getValue());
+
+            determineYoungestMilestoneInQueue();
+        }
+    }
+
+    /**
+     * This method contains the logic for the milestone solidification, that gets executed in a separate {@link Thread}.
+     *
+     * It simply executes the necessary steps periodically while waiting a short time to give the nodes the ability to
+     * answer to the issued transaction requests.
+     */
+    private void milestoneSolidificationThread() {
+        while(!Thread.interrupted()) {
+            processNewlyAddedMilestones();
+            processSolidificationQueue();
+            refillSolidificationQueue();
+
+            ThreadUtils.sleep(SOLIDIFICATION_INTERVAL);
+        }
+    }
+
+    /**
+     * This method processes the newly added milestones.
+     *
+     * We process them lazy to decrease the synchronization requirements and speed up the addition of milestones from
+     * outside {@link Thread}s.
+     *
+     * It simply iterates over the milestones and adds them to the pool. If they are older than the
+     * {@link #youngestMilestoneInQueue}, we add the to the solidification queue.
+     */
+    private void processNewlyAddedMilestones() {
+        for (
+            Iterator<Map.Entry<Hash, Integer>> iterator = newlyAddedMilestones.entrySet().iterator();
+            iterator.hasNext();
+        ) {
+            Map.Entry<Hash, Integer> currentEntry = iterator.next();
+
+            unsolidMilestonesPool.put(currentEntry.getKey(), currentEntry.getValue());
+
+            if (youngestMilestoneInQueue == null || currentEntry.getValue() < youngestMilestoneInQueue.getValue()) {
+                addToSolidificationQueue(currentEntry);
+            }
+
+            iterator.remove();
```

Comment:
no clone required since its a ConcurrentHashMap (see other comment)

Reply:
Thats intended - the next run of processNewlyAddedMilestones() will pick them up so they wont be forgotten. This little "trick" (processing the additions to the queue lazy) allows us to have this class working without any synchronized blocks and make it fast to call.

I actually think that milestones should be analyzed on arrival rather than rescanning all potential milestones over and over again I wanted this class to be as fast as possible. Tbh i think that the milestonetracker and this part of the code is the problem that causes IRI versions to consume more and more memory and CPU over time while the local snapshots version is stable at 7% cpu usage